### PR TITLE
Add dump & restore to holders & simulations.

### LINKED ISSUE #661: Provide a way to dump to/restore fom disk the state of a simulation
Hi there!

I really enjoy OpenFisca, but I recently encountered an issue.

### Here is what I want to do:

I would like to save the state of a simulation  to be able to restart from that exact point several times.
I should thus be able to dump it and reload it.
I think that a good starting point could be the work done by @fpagnoux to allow out-of-memory computations with variables saved on disk.

## Context

I identify more as a:

- Analyst _(I make macroscopic computations on real populations)_.
- Developer _(I create tools that use the existing OpenFisca code)_.

I want to build simulation on huge data and save time at initialisation by always loading the same initial state.
I want to be able to save individual states to run multiple reforms on them.

### PR DESCRIPTION
**Pull request by eraviart**
#### New features

- Add dump & restore to holders & simulations (fix #661).

-------------------

EDIT (@fpagnoux)

Dump:

```py
from openfisca_core.tools.simulation_dumper import  dump_simulation, 
dump_simulation(simulation, '/path/to/directory')
```

Restore:

```py
from openfisca_core.tools.simulation_dumper import restore_simulation
simulation_2 = restore_simulation('/path/to/directory', tax_benefit_system)
```

### CODE REVIEW

**Comment by benjello** [comment]
@fpagnoux : I took the liberty to ask you to review this PR.
@eraviart coded this during a session at IPP.
We need your input before we go forward and we add higher level methods to `SurveyScenario`.

**Comment by fpagnoux** [comment]
Thanks for the PR @eraviart!

A few general questions @benjello before looking more into details:

- With this implem, after we restore the simulation, nothing is loaded in memory and everything is kept on disk. Is that on purpose? This could significantly slow down calculations after restoring. 
- Entities compositions and roles are not stored, is that an issue?

**Comment by benjello** [comment]
@fpagnoux : I knew that your input would be valuable !

- You first point is a very good one. For performance issue we may want that  the variables are kept in RAM when read.
- The second point is another good one, we definitively want to keep those if we want to compute more  variables involving several entities after restoring. I completely forgot that when explaing the problem to @eraviart.

**Comment by fpagnoux** [comment]
Ok, I will try adjust the implem to take these two points into account :)

**Comment by fpagnoux** [comment]
Tested on France calculating `revenu_disponible` for 40000 persons, dumping and restoring takes a reasonable time (<10s)

Mahdi, do you want to try the branch sees if it works for you, or should we go ahead and ask for a review?

**Comment by benjello** [comment]
Thanks @fpagnoux !

Do you confirm that when restoring a simulation, the variables are kept back in RAM/disk according to the new `memroy_config` only when recalled and not restored at once ? 

I think we should ask for a review so I can go on and propagate to downstream packages. 
I could definitively come back and ask for a few changes if needed by a use case.

File: tests/core/test_dump_restore.py [hunk 1]
```diff
@@ -1,141 +1,43 @@
...
-    shutil.rmtree(data_storage_dir)
+from numpy.testing import assert_array_equal
 
+from openfisca_core.simulations import Simulation
+from openfisca_country_template.situation_examples import couple
+from openfisca_core.tools.simulation_dumper import dump_simulation, restore_simulation
 
-def test_dump_restore_different_simulations_different_variables():
-    data_storage_dir = tempfile.mkdtemp(prefix = "openfisca_")
+from .test_countries import tax_benefit_system
```
Comment (sandcha) [inline tests/core/test_dump_restore.py]: Use absolute import?

Comment (fpagnoux) [reply tests/core/test_dump_restore.py]: We can't use absolute imports as the `tests` are not packaged in `openfisca_core`.

**Comment by benjello** [comment]
Looks good to merge then @fpagnoux !

**Pull request merged**